crates/sparrow-runtime/src/execute/output/object_store.rs

```diff
@@ -1,209 +1,263 @@
+use tokio::io::{AsyncWrite, AsyncWriteExt};
 use uuid::Uuid;
 
-use crate::{
-    execute::progress_reporter::ProgressUpdate,
-    s3::{is_s3_path, S3Helper, S3Object},
-};
+use crate::execute::progress_reporter::ProgressUpdate;
+use crate::stores::{ObjectStoreRegistry, ObjectStoreUrl};
+use crate::UPLOAD_BUFFER_SIZE_IN_BYTES;
 
-#[derive(Debug, Display)]
+const ROWS_PER_FILE: usize = 1_000_000;
+
+#[derive(derive_more::Display, Debug)]
 pub enum Error {
-    MalformedUri,
-    ProgressUpdateFailure,
-    UnspecifiedFormat,
-    LocalWriteFailure,
-    UploadFailure,
-    UnsupportedUri,
-    UnsupportedDestination,
+    #[display(fmt = "failed to report output progress")]
+    ProgressUpdate,
+    #[display(fmt = "invalid destination prefix")]
+    InvalidDestination,
+    #[display(fmt = "internal error")]
+    Internal,
+    #[display(fmt = "error starting multi-part upload for output")]
+    Upload,
+    #[display(fmt = "error writing output")]
+    Write,
 }
 
 impl error_stack::Context for Error {}
 
+#[must_use]
+enum WriterState {
+    /// Open Parquet writer.
+    Parquet {
+        url: ObjectStoreUrl,
+        writer: AsyncArrowWriter<Box<dyn AsyncWrite + Send + Unpin>>,
+    },
+    Csv {
+        url: ObjectStoreUrl,
+        writer: Box<dyn AsyncWrite + Send + Unpin>,
+    },
+}
+
+impl WriterState {
+    async fn open(
+        object_store: &dyn ObjectStore,
+        output_prefix: &ObjectStoreUrl,
+        file_type: FileType,
+        schema: SchemaRef,
+    ) -> error_stack::Result<Self, Error> {
+        let url = output_prefix
+            .join(&format!("{}", Uuid::new_v4().as_hyphenated()))
+            .change_context(Error::Internal)?;
+        let path = url.path().change_context(Error::Internal)?;
+        let (_multipart_id, mut writer) = object_store
+            .put_multipart(&path)
+            .await
+            .into_report()
+            .change_context(Error::Upload)?;
+
+        match file_type {
+            FileType::Unspecified => todo!(),
+            FileType::Parquet => {
+                let writer = parquet::arrow::AsyncArrowWriter::try_new(
+                    writer,
+                    schema,
+                    UPLOAD_BUFFER_SIZE_IN_BYTES,
+                    None,
+                )
+                .into_report()
+                .change_context(Error::Write)?;
+                Ok(Self::Parquet { url, writer })
+            }
+            FileType::Csv => {
+                // Write an empty batch so headers are written.
+                let mut buffer = Vec::new();
+                let mut csv_writer = arrow::csv::WriterBuilder::new()
+                    .has_headers(true)
+                    .build(&mut buffer);
+                csv_writer
+                    .write(&RecordBatch::new_empty(schema))
+                    .into_report()
+                    .change_context(Error::Write)?;
+                std::mem::drop(csv_writer);
```

Comment:
you mentioned this in a comment recently - prefer the explicit drop, or can you place the `csv_writer` in its own `{ .. } ` block?

Reply:
The issue here was that the `csv_writer` takes `&mut buffer`, so until it is dropped we can't get `&buffer` to write to the actual writer. I think I'm fine with it's own block hree to explicitly scope the writer. Thoughts?